sonyc_kalman/data.py

```diff
@@ -70,6 +70,93 @@ def load_openl3_time_series(hdf5_path, delta_mins=15, aggr_func=None):
     return X, mask
 
 
+def series_splitter(invalid_mask, test_ratio=0.25, mode='random_holes', hole_param=[20, 10]):
+    '''
+    Given a `invalid_mask` of length n, generate masks for the train split and test split according to `test_ratio`.
+
+    Params:
+    -------
+        invalid_mask: list or np.array
+            an array of 1's and 0's. 0 being valid, 1 being invalid
+        test_ratio: float (0, 1)
+            the fraction of the data to be split out as test
+        mode: str
+            should be one of 'random_holes', 'r', 'chronological', or 'c'.
+            'random_holes' - punch_out randomly (Guasssian according to `hole_param`) sized holes in the series and 
+            set aside as test.
+            'chronological' - just use the begining as train and the end as test, modeling real world delopyment.
+            hole_param in chronological mode is ignored.
+        hole_param: list [mean, std]
+            used for the 'random_holes' mode.
+            
+    Returns:
+    --------
+        train_mask: np.array of int's
+            an array of 1's and 0's that can be used to mask any series. 1 means selected for the training split
+        test_mask: np.array of int's
+            an array of 1's and 0's that can be used to mask any series. 1 means selected for the test split
+    '''
+    if mode == 'chronological' or mode == 'c':
+        chrono_mode = True
+    elif mode == 'random_holes' or mode == 'r':
+        chrono_mode = False
+    else:
+        raise ValueError("type must be one of 'random_holes', 'r', 'chronological', or 'c'.")
+
+    invalid_mask = np.array(invalid_mask, dtype=np.bool)
+    valid_mask = 1 - invalid_mask
+    valid_mask = valid_mask.astype(np.bool)
+    n_total_valid = sum(valid_mask)
+    n_test = int(round(test_ratio * n_total_valid))
+    n_train = n_total_valid - n_test
+    
+    train_mask = valid_mask.copy()
+    test_mask = valid_mask.copy()
+    if chrono_mode:
+        # first split the data as if all the valid frames are contiguous
+        continuous_split_test = np.array([0] * n_train + [1] * n_test, dtype=int)
+        continuous_split_train = 1 - continuous_split_test
+        # then use the valid_mask to put these contiguous masks into the right places
+        test_mask[valid_mask]  = continuous_split_test
```

Comment:
Note that we can still use the invalid samples in the test/training set, just that they won't be used as observations. We should still include them in the train/test splits.

Reply:
Ok, so for all the invalid ones, I won't mask them out. In other words, if `mask[i] == 1` then 'test_mask[i] == 0, train_mask[i] == 0` right?

Reply:
I think what I meant was that the mask doesn't need to factor into this at all. All samples (including invalid samples) should be included in the train/test split.

A related concern: after thinking about it a bit more, _ideally_ we should only have invalid samples in the training set, and none in the test set. Though, this doesn't fundamentally change how we evaluate. We can still have invalid samples in the test set and then just measure error over valid samples.

We could also deal with boundary conditions, i.e. ensuring that we don't start or end a sequence with invalid samples. But we can punt this to later.

Reply:
> ideally we should only have invalid samples in the training set
This is how I'll do it. since the ratio reflect how much valid data is held-out for the test/hole partition